scio-redis/src/main/scala/com/spotify/scio/redis/RedisDoFn.scala

```diff
@@ -0,0 +1,137 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.scio.redis
+
+import org.apache.beam.sdk.io.redis.RedisConnectionConfiguration
+import org.apache.beam.sdk.transforms.DoFn._
+import org.apache.beam.sdk.transforms.DoFn
+import redis.clients.jedis.{Jedis, Pipeline}
+import redis.clients.jedis.Response
+import java.util.concurrent.ConcurrentLinkedQueue
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow
+import org.apache.beam.sdk.transforms.display.DisplayData.Builder
+import org.apache.beam.sdk.transforms.display.DisplayData
+import org.joda.time.Instant
+import scala.concurrent.{Future, Promise}
+import scala.util._
+import scala.concurrent.{Await, ExecutionContext}
+import scala.concurrent.duration.Duration
+import scala.jdk.CollectionConverters._
+
+abstract class RedisDoFn[I, O](
+  connectionConfig: RedisConnectionConfiguration,
+  batchSize: Int
+)(implicit ec: ExecutionContext)
+    extends DoFn[I, O] {
+
+  @transient private var jedis: Jedis = _
+  @transient private var pipeline: Pipeline = _
+  private val results: ConcurrentLinkedQueue[Future[Result]] = new ConcurrentLinkedQueue()
+  private val requests: ConcurrentLinkedQueue[(List[Response[_]], Promise[List[_]])] =
+    new ConcurrentLinkedQueue()
+  private var batchCount = 0
+  private val client = new Client()
+
+  private case class Result(input: I, output: O, ts: Instant, w: BoundedWindow)
+
+  abstract class Request {
+    def create(pipeline: Pipeline): List[Response[_]]
+  }
+
+  final class Client extends Serializable {
+    def request(request: Request): Future[List[_]] = {
+      val promise = Promise[List[_]]()
+      requests.add((request.create(pipeline), promise))
+      promise.future
+    }
+  }
+
+  def this(opts: RedisConnectionOptions, batchSize: Int)(implicit ec: ExecutionContext) =
+    this(RedisConnectionOptions.toConnectionConfig(opts), batchSize)
+
+  private def flush(fn: Result => Unit): Unit = {
+    pipeline.exec
+    pipeline.sync()
+
+    val iter = requests.iterator()
+    while (iter.hasNext()) {
+      val (rsp, promise) = iter.next()
+      promise.success(rsp.flatMap(r => Option(r.get())))
+    }
+
+    val future = Future.sequence(results.asScala).andThen {
+      case Success(value) =>
+        val iter = value.iterator
+        while (iter.hasNext) {
+          fn(iter.next())
+        }
+      case Failure(_) => ()
+    }
+
+    Await.result(future, Duration.Inf)
+
+    results.clear()
+    requests.clear()
+  }
+
+  def request(value: I, client: Client): Future[O]
+
+  @Setup
+  def setup(): Unit =
+    jedis = connectionConfig.connect
+
+  @StartBundle
+  def startBundle(): Unit = {
+    pipeline = jedis.pipelined
+    pipeline.multi
+    batchCount = 0
+  }
+
+  @ProcessElement
+  def processElement(c: ProcessContext, window: BoundedWindow): Unit = {
+    val result = request(c.element(), client).map { r =>
```

Comment:
info: so technically all Redis requests will be put into the Jedis pipeline in the current thread. The only place where it forks is in `RedisWriter.Writer` in the `map(_ => ())`. `pipeline.exec/sync` sends all requests and reads everything back in the current thread as well. The only concurrency we have here is after all Redis requests are completed. I am wondering what we gain in return by adding future/promises here. Is the main idea to have concurrency in the part processing Redis results? (Might be a dumb question - might be missing the main point here)

Reply:
The reason to use `Future`/`Promise` is merely to convey to the user that the result from the `Request` function is not computed right away as it is included in the pipeline bundle. Without it users would be able to try getting a value out of a response that was not yet obtained.

Reply:
Got it, this makes sense. Looks like clients still can try to access result before it was processed:

```scala
new RedisDoFn[T, Unit](connectionConfig, batchSize) {
        override def request(value: T, client: Client): Future[Unit] =
          client
            .request { pipeline =>
                val resps = RedisMutator.mutate(pipeline)(value)
                // do something with resps...
                resps
             }
            .map(_ => ())
      }
```

But probably it is difficult to put a guard against this.

Reply:
Yeah they can, but that's fine as they will not be able to do much.